GenMat.cpp

```diff
@@ -0,0 +1,544 @@
+  C.elements[row * C.width + col] = cValue;
+}
+
+
+__global__ void MatByFloat(GMat A, float b, GMat C)
+{
+   int row = blockIdx.y * blockDim.y + threadIdx.y;
+   int col = blockIdx.x * blockDim.x + threadIdx.x;
+   C.elements[row * C.width + col] = A.elements[row * A.width +col] * b;
+
+}
+
+__global__ void MatByDouble(GMat A, double b, GMat C)
+{
+   int row = blockIdx.y * blockDim.y + threadIdx.y;
+   int col = blockIdx.x * blockDim.x + threadIdx.x;
+   C.elements[row * C.width + col] = A.elements[row * A.width +col] * b;
+
+}
+
+__global__ void MatAdd(GMat A, GMat B, GMat C)
+{
+   int row = blockIdx.y * blockDim.y + threadIdx.y;
+   int col = blockIdx.x * blockDim.x + threadIdx.x;
+   C.elements[row * C.width + col] = A.elements[row * A.width + col] + B.elements[row * B.width + col];
+
+}
+
+__global__ void Transpose(GMat A, GMat C)
+{
+   int row = blockIdx.y * blockDim.y + threadIdx.y;
+   int col = blockIdx.x * blockDim.x + threadIdx.x;
+   C.elements[row * C.width + col] =  A.elements[col * A.width + row];
+}
+
+__global__ void GpuCopy(GMat A, GMat C)
+{
+   int row = blockIdx.y * blockDim.y + threadIdx.y;
+   int col = blockIdx.x * blockDim.x + threadIdx.x;
+   C.elements[row * C.width + col] =  A.elements[row * A.width + col];
+}
+
+void GEMM(const int alpha, const bool opA, const CMat A, const bool opB, const CMat B, const int beta, const CMat C, CMat D)
+{
+
+
+  // Load A, B, and C to device memory
+  // Allocate Matrix D
+  GMat dA(A.width, A.height);
+  GMat dB(B.width, B.height);
+  GMat dC(C.width, C.height);
+  GMat dD(D.width, D.height);
+
+
+  //CMat aT(A.height,A.width);
+  //GMat dAt(A.height,A.width);
+
+  CMat addResult(C.width, C.height);
+  GMat dAddResult(C.width, C.height);
+
+  CMat AbyB(A.height, B.height);
+  GMat dAbyB(A.height, B.height);
+
+  CMat CbyBeta(C.width, C.height);
+  GMat dCbyBeta(C.width, C.height);
+
+
+  dA.load(A);
+  dB.load(B);
+  dC.load(C);
+
+
+  int aTransHeight;
+  int aTransWidth;
+  if(opA)
+  {
+     aTransHeight = A.width;
+     aTransWidth = A.height;
+  }
+  else
+  {
+     aTransHeight = A.height;
+     aTransWidth = A.width;
+  }
+
+  int bTransHeight;
+  int bTransWidth;
+  if(opB)
+  {
+     bTransHeight = B.width;
+     bTransWidth = B.height;
+  }
+  else
+  {
+     bTransHeight = B.height;
+     bTransWidth = B.width;
+  }
+
+  CMat aT(aTransWidth,aTransHeight);
+  GMat dAt(aTransWidth,aTransHeight);
+
+  CMat bT(bTransWidth,bTransHeight);
+  GMat dBt(bTransWidth,bTransHeight);
+
+ 
+  // D = alpha * opA (A) * opB (B) + beta(C)
+
+  // A transpose
+  if(opA)
+  {
+      for (int i = 0; i < A.height; i++) 
+      {
+       for (int j = 0; j < A.width; j++) {
+        std::cout << A.elements[i * A.width + j];
+        std::cout << " ";
+      }
+      std::cout << std::endl;
+     }
+     std::cout << std::endl;
+     std::cout << std::endl;
+
+
+     dim3 dimBlockTransA(gcd(aT.width,aT.height),gcd(aT.width,aT.height));
```

Comment:
The block dimension is generally a tunable parameter. I'm not positive using the gcd is really how you want to do this. Also, there's no guarantee here that `dimBlockTransA.x * dimBlockTransA.y <= 1024`